File: IAIso-v5.0/core/iaiso-swift/Tests/IAIsoIdentityTests/IdentityTests.swift
import XCTest
@testable import IAIsoConsent
@testable import IAIsoIdentity

final class IdentityTests: XCTestCase {

    func testDeriveDirectClaimString() {
        let out = OidcVerifier.deriveScopes(
            ["scope": "tools.search tools.fetch"],
            mapping: .defaults)
        XCTAssertTrue(out.contains("tools.search"))
        XCTAssertTrue(out.contains("tools.fetch"))
    }

    func testDeriveDirectClaimArray() {
        let out = OidcVerifier.deriveScopes(
            ["scp": ["a.b", "c"]],
            mapping: .defaults)
        XCTAssertTrue(out.contains("a.b"))
        XCTAssertTrue(out.contains("c"))
    }

    func testDeriveGroupToScopes() {
        let out = OidcVerifier.deriveScopes(
            ["groups": ["engineers"]],
            mapping: ScopeMapping(
                groupToScopes: ["engineers": ["tools.search", "tools.fetch"]]))
        XCTAssertTrue(out.contains("tools.search"))
        XCTAssertTrue(out.contains("tools.fetch"))
    }

    func testAlwaysGrantAdded() {
        let out = OidcVerifier.deriveScopes(
            [:],
            mapping: ScopeMapping(alwaysGrant: ["base"]))
        XCTAssertEqual(out, ["base"])
    }

    func testPresetsHaveExpectedEndpoints() {
        let okta = ProviderConfig.okta(domain: "acme.okta.com", audience: "api")
        XCTAssertTrue(okta.discoveryURL!.contains("acme.okta.com"))

        let auth0 = ProviderConfig.auth0(domain: "acme.auth0.com", audience: "api")
        XCTAssertTrue(auth0.issuer!.hasSuffix("/"))

        let azure = ProviderConfig.azureAd(tenant: "tenant-id", audience: "api", v2: true)
        XCTAssertTrue(azure.discoveryURL!.contains("v2.0"))
    }

    func testVerifyFailsWhenJwksNotLoaded() {
        let v = OidcVerifier(config: .defaults)
        XCTAssertThrowsError(try v.verify("a.b.c"))
    }
}

File: IAIso-v5.0/core/iaiso-csharp/tests/Iaiso.Tests/IdentityTests.cs
using System.Collections.Generic;
using System.Text.Json.Nodes;
using Iaiso.Identity;

namespace Iaiso.Tests;

public sealed class IdentityTests
{
    public void TestDeriveDirectClaimString()
    {
        var claims = JsonNode.Parse("{\"scope\": \"tools.search tools.fetch\"}")!.AsObject();
        var @out = OidcVerifier.DeriveScopes(claims, ScopeMapping.Defaults());
        Assert.True(@out.Contains("tools.search"));
        Assert.True(@out.Contains("tools.fetch"));
    }

    public void TestDeriveDirectClaimArray()
    {
        var claims = JsonNode.Parse("{\"scp\": [\"a.b\", \"c\"]}")!.AsObject();
        var @out = OidcVerifier.DeriveScopes(claims, ScopeMapping.Defaults());
        Assert.SequenceEqual(new[] { "a.b", "c" }, @out);
    }

    public void TestDeriveGroupToScopes()
    {
        var claims = JsonNode.Parse("{\"groups\": [\"engineers\"]}")!.AsObject();
        var g = new Dictionary<string, IReadOnlyList<string>>
        {
            ["engineers"] = new[] { "tools.search", "tools.fetch" },
        };
        var mapping = new ScopeMapping(new List<string>(), g, null);
        var @out = OidcVerifier.DeriveScopes(claims, mapping);
        Assert.True(@out.Contains("tools.search"));
        Assert.True(@out.Contains("tools.fetch"));
    }

    public void TestAlwaysGrantAdded()
    {
        var claims = new JsonObject();
        var mapping = new ScopeMapping(new List<string>(), null, new[] { "base" });
        Assert.SequenceEqual(new[] { "base" }, OidcVerifier.DeriveScopes(claims, mapping));
    }

    public void TestPresetsHaveExpectedEndpoints()
    {
        var okta = ProviderConfig.Okta("acme.okta.com", "api");
        Assert.Contains("acme.okta.com", okta.DiscoveryUrl!);
        var auth0 = ProviderConfig.Auth0("acme.auth0.com", "api");
        Assert.True(auth0.Issuer!.EndsWith("/"));
        var azureV2 = ProviderConfig.AzureAd("tenant-id", "api", true);
        Assert.Contains("v2.0", azureV2.DiscoveryUrl!);
    }

    public void TestVerifyFailsWhenJwksNotLoaded()
    {
        var v = new OidcVerifier(ProviderConfig.Defaults());
        Assert.Throws<IdentityException>(() => v.Verify("a.b.c"));
    }
}
